File: packages/react-native-notificare-in-app-messaging/ios/NotificareInAppMessagingModule.swift
import React
import NotificareKit
import NotificareInAppMessagingKit

private let DEFAULT_ERROR_CODE = "notificare_error"

@objc(NotificareInAppMessagingModule)
class NotificareInAppMessagingModule: RCTEventEmitter {

    private var hasListeners = false
    private var eventQueue = [(name: String, payload: Any?)]()

    override init() {
        super.init()

        Notificare.shared.inAppMessaging().delegate = self
    }

    override class func requiresMainQueueSetup() -> Bool {
        return true
    }

    override func startObserving() {
        hasListeners = true

        if !eventQueue.isEmpty {
            NotificareLogger.debug("Processing event queue with \(eventQueue.count) items.")
            eventQueue.forEach { sendEvent(withName: $0.name, body: $0.payload)}
            eventQueue.removeAll()
        }
    }

    override func stopObserving() {
        hasListeners = false
    }

    override func supportedEvents() -> [String] {
        return [
            "re.notifica.iam.message_presented",
            "re.notifica.iam.message_finished_presenting",
            "re.notifica.iam.message_failed_to_present",
            "re.notifica.iam.action_executed",
            "re.notifica.iam.action_failed_to_execute",
        ]
    }

    private func dispatchEvent(_ name: String, payload: Any?) {
        if hasListeners {
            sendEvent(withName: name, body: payload)
        } else {
            eventQueue.append((name: name, payload: payload))
        }
    }

    // MARK: - Notificare In-App Messaging

    @objc func hasMessagesSuppressed(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        resolve(Notificare.shared.inAppMessaging().hasMessagesSuppressed)
    }
    
    @objc func setMessagesSuppressed(_ data: [String: Any], resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        guard let suppressed = data["suppressed"] as? Bool else {
            reject(DEFAULT_ERROR_CODE, "Missing 'suppressed' parameter.", nil)
            return
        }
        
        let evaluateContext = data["evaluateContext"] as? Bool ?? false
        
        Notificare.shared.inAppMessaging().setMessagesSuppressed(suppressed, evaluateContext: evaluateContext)
        
        resolve(nil)
    }
}

extension NotificareInAppMessagingModule: NotificareInAppMessagingDelegate {
    func notificare(_ notificare: NotificareInAppMessaging, didPresentMessage message: NotificareInAppMessage) {
        do {
            dispatchEvent("re.notifica.iam.message_presented", payload: try message.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.iam.message_presented event.", error: error)
        }
    }

    func notificare(_ notificare: NotificareInAppMessaging, didFinishPresentingMessage message: NotificareInAppMessage) {
        do {
            dispatchEvent("re.notifica.iam.message_finished_presenting", payload: try message.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.iam.message_finished_presenting event.", error: error)
        }
    }

    func notificare(_ notificare: NotificareInAppMessaging, didFailToPresentMessage message: NotificareInAppMessage) {
        do {
            dispatchEvent("re.notifica.iam.message_failed_to_present", payload: try message.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.iam.message_failed_to_present event.", error: error)
        }
    }

    func notificare(_ notificare: NotificareInAppMessaging, didExecuteAction action: NotificareInAppMessage.Action, for message: NotificareInAppMessage) {
        do {
            dispatchEvent("re.notifica.iam.action_executed", payload: [
                "message": try message.toJson(),
                "action": try action.toJson(),
            ])
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.iam.action_executed event.", error: error)
        }
    }

    func notificare(_ notificare: NotificareInAppMessaging, didFailToExecuteAction action: NotificareInAppMessage.Action, for message: NotificareInAppMessage, error: Error?) {
        do {
            var payload: [String: Any] = [
                "message": try message.toJson(),
                "action": try action.toJson(),
            ]

            if let error = error {
                payload["error"] = error.localizedDescription
            }

            dispatchEvent("re.notifica.iam.action_failed_to_execute", payload: payload)
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.iam.action_failed_to_execute event.", error: error)
        }
    }
}

File: packages/react-native-notificare-in-app-messaging/android/src/main/java/re/notifica/iam/react_native/NotificareInAppMessagingModule.kt
package re.notifica.iam.react_native

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import re.notifica.Notificare
import re.notifica.iam.NotificareInAppMessaging
import re.notifica.iam.ktx.inAppMessaging
import re.notifica.iam.models.NotificareInAppMessage
import re.notifica.internal.NotificareLogger

public class NotificareInAppMessagingModule internal constructor(context: ReactApplicationContext) :
    NotificareInAppMessagingModuleSpec(context), NotificareInAppMessaging.MessageLifecycleListener {

    override fun getName(): String {
        return NAME
    }

    override fun initialize() {
        super.initialize()

        EventBroker.setup(reactApplicationContext)
        Notificare.inAppMessaging().addLifecycleListener(this)
    }

    override fun invalidate() {
        super.invalidate()

        Notificare.inAppMessaging().removeLifecycleListener(this)
    }

    @ReactMethod
    override fun addListener(eventName: String) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    override fun removeListeners(count: Double) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    // region Notificare In-App Messaging

    @ReactMethod
    override fun hasMessagesSuppressed(promise: Promise) {
        promise.resolve(Notificare.inAppMessaging().hasMessagesSuppressed)
    }

    @ReactMethod
    override fun setMessagesSuppressed(data: ReadableMap, promise: Promise) {
        val arguments = data.toJson()
        val suppressed = try {
            arguments.getBoolean("suppressed")
        } catch (e: Exception) {
            promise.reject(DEFAULT_ERROR_CODE, e)
            return
        }

        val evaluateContext =
            if (!arguments.isNull("evaluateContext")) {
                arguments.getBoolean("evaluateContext")
            } else {
                false
            }

        Notificare.inAppMessaging().setMessagesSuppressed(suppressed, evaluateContext)

        promise.resolve(null)
    }

    // endregion

    // region NotificareInAppMessaging.MessageLifecycleListener

    override fun onMessagePresented(message: NotificareInAppMessage) {
        try {
            EventBroker.dispatchEvent("re.notifica.iam.message_presented", message.toJson().toReactMap())
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the re.notifica.iam.message_presented event.", e)
        }
    }

    override fun onMessageFinishedPresenting(message: NotificareInAppMessage) {
        try {
            EventBroker.dispatchEvent("re.notifica.iam.message_finished_presenting", message.toJson().toReactMap())
        } catch (e: Exception) {
            NotificareLogger.error("re.notifica.iam.Failed to emit the message_finished_presenting event.", e)
        }
    }

    override fun onMessageFailedToPresent(message: NotificareInAppMessage) {
        try {
            EventBroker.dispatchEvent("re.notifica.iam.message_failed_to_present", message.toJson().toReactMap())
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the re.notifica.iam.message_failed_to_present event.", e)
        }
    }

    override fun onActionExecuted(message: NotificareInAppMessage, action: NotificareInAppMessage.Action) {
        try {
            val arguments = Arguments.createMap()
            arguments.putMap("message", message.toJson().toReactMap())
            arguments.putMap("action", message.toJson().toReactMap())

            EventBroker.dispatchEvent("re.notifica.iam.action_executed", arguments)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the re.notifica.iam.action_executed event.", e)
        }
    }

    override fun onActionFailedToExecute(
        message: NotificareInAppMessage,
        action: NotificareInAppMessage.Action,
        error: Exception?
    ) {
        try {
            val arguments = Arguments.createMap()
            arguments.putMap("message", message.toJson().toReactMap())
            arguments.putMap("action", message.toJson().toReactMap())

            if (error != null) {
                arguments.putString("error", error.message)
            }

            EventBroker.dispatchEvent("re.notifica.iam.action_failed_to_execute", arguments)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the re.notifica.iam.action_failed_to_execute event.", e)
        }
    }

    // endregion

    public companion object {
        internal const val NAME = "NotificareInAppMessagingModule"
        internal const val DEFAULT_ERROR_CODE = "notificare_error"
    }
}
